docs/source/ops/monitoring.rst

```diff
@@ -0,0 +1,130 @@
+Monitoring
+==========
+
+This section gives a high-level overview of the PrivateStorageIO monitoring efforts.
+
+
+Goals
+`````
+
+Alerting
+  Something might break soon, so somebody needs to do something.
+
+Comparing over time or experiment groups
+  Is our service slower than it was last week? Does database B answer queries faster than database A?
+
+Analyzing long-term trends
+  How big is my database and how fast is it growing? How quickly is my daily-active user count growing?
+
+
+Introduction to our dashboards
+``````````````````````````````
+
+We have two groups of dashboards: Requests (external view, RED method) and Resources (internal view, USE method).
+
+Services and their dependencies can be visualized as a tree from external-facing to internal systems.
+
+.. figure:: DAG-of-services.svg
+    :width: 200px
+    :align: center
+    :alt: DAG of services, created using https://www.nomnoml.com/
+    :figclass: align-center
+
+    DAG of services
+
+We order our dashboards like a breadth-first-search of that tree.
+This makes it easier to understand dependencies and faster to trouble shoot when a high-latency problem on a low-level service bubbles up.
+
+
+Meaning of our metrics
+``````````````````````
+
+Google's *Monitoring Distributed Systems* book about what they call the `Four Golden Signals <https://sre.google/sre-book/monitoring-distributed-systems/#xref_monitoring_golden-signals>`_ has a great explanation and definition of useful metrics:
+
+Latency
+  Requests, but also errors take time, so don't discard them.
+
+Traffic
+  What constitutes "Traffic" depends on the nature of your system.
+
+Errors
+  (The *rate* of ) failed requests.
+
+Saturation
+  How "full" your service is.  Take action (i.e. page a human) before service degrades.
+
+"If you measure all four golden signals and page a human when one signal is problematic (or, in the case of saturation, nearly problematic), your service will be at least decently covered by monitoring."
+
+
+RED method for services ("request-scoped", "external view")
+```````````````````````````````````````````````````````````
+
+Request rate, Errors, Duration (+ Saturation?)
+
+* Instrument everything that takes time and could fail
+   * "In contrast to logging, services should instrument every meaningful number available for capture." (Peter Bourgon)
+
+* Plot 99th percentile, 50th percentile and average
+   * 50th and average should be close - else something is wrong
+   * Averages sum neatly - Service latency average should be sum of child service latencies
+
+
+USE method for resources ("resource-scoped", "internal view")
+`````````````````````````````````````````````````````````````
+
+Utilization, Saturation, Errors:
+
+* CPUs
+   * Saturation (Per core? So we catch maxxed out single cores? Would that work?)
+* Memory
+   * Capacity
+* Network
+* Disks
+   * Capacity
+   * I/O
+* Interconnects (Really? Hard to measure & shouldn't be the problem most of the time...)

```

Comment:
I'm not sure what this one means.  Connections between servers?   I see Google has a thing they called interconnects - https://cloud.google.com/monitoring/api/metrics_gcp#gcp-interconnect - which looks like what I would usually call "network interfaces" but maybe their name is better.

Reply:
Internal system buses. This comes from the [original USE metrics](http://www.brendangregg.com/USEmethod/use-linux.html), where Brendan Gregg reasons:

> CPU, memory and I/O interconnects are often overlooked. Fortunately, they aren't commonly the system bottleneck. Unfortunately, if they are, it can be difficult to do much about (maybe you can upgrade the main board, or reduce load: eg, "zero copy" projects lighten memory bus load).

Examples: My last software project (the pedestrian simulator) was heavily dependent on system memory throughput. The NVMe SSDs of the Let's Encrypt main database server I mentioned in the "#random" slack channel can max out the PCI bus. 

Let's revisit this later if we need to.